secret/store.go

```diff
@@ -2,26 +2,81 @@ package secret
 
 import (
 	"context"
+	"sync"
 
 	"github.com/dagger/dagger/core"
 	bkgw "github.com/moby/buildkit/frontend/gateway/client"
 	"github.com/moby/buildkit/session/secrets"
 )
 
 func NewStore() *Store {
-	return &Store{}
+	return &Store{
+		nameToDigest:  map[string]string{},
+		idToPlaintext: map[core.SecretID]string{},
+	}
 }
 
 var _ secrets.SecretStore = &Store{}
 
 type Store struct {
 	gw bkgw.Client
+
+	mu            sync.Mutex
+	nameToDigest  map[string]string
+	idToPlaintext map[core.SecretID]string
 }
 
 func (store *Store) SetGateway(gw bkgw.Client) {
 	store.gw = gw
 }
 
+// AddSecret adds the secret identified by user defined name with its plaintext
+// value to the secret store.
+func (store *Store) AddSecret(_ context.Context, name, plaintext string) (core.SecretID, error) {
+	store.mu.Lock()
+	defer store.mu.Unlock()
+
+	id, err := core.NewSecretID(name, plaintext)
+	if err != nil {
+		return id, err
+	}
+
+	digest, err := id.Digest()
+	if err != nil {
+		// We shouldn't arrive here unless we messed up on our side.
+		// We should stop here to analyze what's wrong.
+		panic(err)
+	}
```

Comment:
It looks like we only use `NewSecretID` in this call site, and the secret ID itself only stores the digest. Computing the digest can't fail, and we need it outside of `NewSecretID`, so maybe we should just compute it outside and pass the digest in? That would eliminate this impossible err case.

Reply:
I completely removed the `Digest()` method as it became useless.
It is computed before hand, added in the SecretID. Then everything happens thanks to the SecretID, and I don't need to have a dedicated name to digest anymore, as the ID is the name+digest already. It clearly showed re-reading my code, the nameToDigest map was never read…

Reply:
done in [9f34bcf](https://github.com/dagger/dagger/pull/4667/commits/9f34bcfebd66df659bdbe83898e6da2b96a80fa6)